### PR TITLE
perf: Speedup ndjson reader `~40%`

### PR DESCRIPTION
**Pull request by ChayimFriedman2**
This speeds it up by 40% in the following benchmark:
```rust
use criterion::{criterion_group, criterion_main, Criterion};
use mimalloc::MiMalloc;
use polars::prelude::*;
use polars_core::POOL;
use rand::{thread_rng, Rng};

#[global_allocator]
static MIMALLOC: MiMalloc = MiMalloc;

fn from_json(json: &[u8]) -> DataFrame {
    JsonReader::new(std::io::Cursor::new(json))
        .with_json_format(JsonFormat::JsonLines)
        .set_rechunk(false)
        .finish()
        .unwrap()
}

fn my_benchmark(_c: &mut Criterion) {
    POOL.install(|| {
        let mut c = Criterion::default().configure_from_args();

        const SIZE: i32 = 5_000_000;
        let mut rng = thread_rng();
        let mut df = df![
            "a" => (0..SIZE).map(|_| rng.gen::<i32>()).collect::<Vec<_>>(),
            "b" => (0..SIZE).map(|v| v.to_string()).collect::<Vec<_>>(),
        ]
        .unwrap();

        let mut json = Vec::new();
        JsonWriter::new(&mut json)
            .with_json_format(JsonFormat::JsonLines)
            .finish(&mut df)
            .unwrap();

        c.bench_function("JSON Lines Deserialization", |b| {
            b.iter(|| from_json(&json))
        });
    });
}

criterion_group!(benches, my_benchmark);
criterion_main!(benches);
```

I had a plan to improve it more, but I can't find time for that and this will involve bigger changes, even a rewrite of the mechanism, while the changes in this PR are simple and effective, so I thought I'll just send them.

Best reviewed commit-by-commit.

A warning from the second commit, repeated here for noticeability:

> This *could* break people's code since we will not split correctly (and thus error) if one object spans two lines or two objects are in the same line. However, such code was already broken, since NDJSON is not allowed to contain any line breaks. If this is a concern, it is possible (at some perf degradation) to check for `}\n` instead of `\n` alone, and that will make this basically equivalent to the splitting logic we have for threads.

### CODE REVIEW

File: crates/polars-io/src/ndjson/core.rs [hunk 1]
```diff
@@ -406,23 +405,25 @@ struct Scratch {
     buffers: simd_json::Buffers,
 }
 
+fn json_lines(bytes: &[u8]) -> impl Iterator<Item = &[u8]> {
+    // This previously used `serde_json`'s `RawValue` to deserialize chunks without really deserializing them.
+    // However, this convenience comes at a cost. serde_json allocates and parses and does UTF-8 validation, all
+    // things we don't need since we use simd_json for them. Also, `serde_json::StreamDeserializer` has a more
+    // ambitious goal: it wants to parse potentially *non-delimited* sequences of JSON values, while we know
+    // our values are line-delimited. Turns out, custom splitting is very easy, and gives a very nice performance boost.
+    bytes.split(|&byte| byte == b'\n').filter(|&bytes| {
```
Comment (ritchie46) [inline crates/polars-io/src/ndjson/core.rs:403]: Does `NDJSON` not support lines delimiter within string fields? E.g.

`{"My text with a line split.\nAnd this is the second line."}`.


If it does, we could also reuse `split_lines` iterator from the csv logic, which checks that properly.

**Review by ritchie46 — COMMENTED** [review]
Really nice PR and great review with those commit messages. I have one question.

File: crates/polars-io/src/ndjson/core.rs (hunk 1 above)
Comment (ChayimFriedman2) [reply crates/polars-io/src/ndjson/core.rs:403]: Per [the specification](https://github.com/ndjson/ndjson-spec?tab=readme-ov-file#31-serialization), this is incorrect (newlines in strings needs to be escaped). Of course, that doesn't mean users are not relying on this, which may mean we want to support that too :)

Comment (ChayimFriedman2) [reply crates/polars-io/src/ndjson/core.rs:403]: Also, the current line split for threads does not check this (it only checks for `}\n`), which means that users relying on this have a subtle bug that may manifest only occasionally.

Comment (ritchie46) [reply crates/polars-io/src/ndjson/core.rs:403]: Ok, then leave it like this. I want us to follow specs (or the closest thing there is to a spec, looking at csv).

**Pull request merged**